src/borrowing.rs

```diff
@@ -0,0 +1,109 @@
+//! Non-allcating, safe, statically allocated wrapper for `DiatomicWaker`.
+//!
+//! See the [crate-level documentation](crate) for usage.
+
+use std::task::Waker;
+
+use crate::primitives::DiatomicWaker;
+use crate::primitives::WaitUntil;
+
+/// A safe, statically allocated `DiatomicWaker`
+///
+/// See the [crate-level documentation](crate) for usage.
+#[derive(Default, Debug)]
+pub struct BorrowingDiatomicWaker {
+    inner: DiatomicWaker,
+}
+
+/// An object that can await a notification from one or several
+/// [`WakeSource`](WakeSource)s.
+///
+/// See the [crate-level documentation](crate) for usage.
+#[derive(Debug)]
+pub struct WakeSink<'a> {
```

Comment:
To prevent confusion with the other `WakeSink`, I would prefer to use another name such as `WakeSinkRef`.